### PR TITLE
[DRAFT] Restructuring the code #383

### PR DESCRIPTION
**Pull request by Jeadie**
Changing code structure for #383 

Still in draft, fielding consenus/feedback.

Further refactors will be, but not included in this:
 - Refactor `config.yml` to have a `default` tag for common params.
 - Use `FROM ann-benchmarks-abc` for algorithm XYZ when applicable

### CODE REVIEW

**Comment by erikbern** [comment]
This is awesome! Let me know how we can get this merged. Think this structure is a lot cleaner.

**Comment by Jeadie** [comment]
No problem. I will try getting this ready for merge this week (to avoid ongoing merge conflicts). Currently working on this. Right now what is left to do it just minimally reimplement `definition.py`. I think there's better future refactors for this, but I want this PR to provide the least amount of logic changes.

**Comment by Jeadie** [comment]
Okay, changes are up. I have had a few problems with some of the tests in `benchmarks.yml`. All except for sptag are transient, that is they have at least one (generally more) passing attempts in my action history (once completing the featureset, of course). SPTAG seems to be the one consistently to fail.

File: .github/workflows/benchmarks.yml [hunk 1]
```diff
@@ -5,6 +5,7 @@ on:
     branches:
       - main
   pull_request:
+  workflow_dispatch:
```
Comment (Jeadie) [inline .github/workflows/benchmarks.yml:8]: I believe `workflow_dispatch` is only allowed for write accessed users (i.e. safe for open-source repo) [[link](https://github.com/orgs/community/discussions/26622)].

**Comment by Jeadie** [comment]
@erikbern fixes in for restructure. Hopefully we can get it merged before more conflicts.

GH action [run](https://github.com/Jeadie/ann-benchmarks/actions/runs/5196127060/jobs/9369527314). SPTAG failing for build problems.

**Comment by erikbern** [comment]
This looks AMAZING!! I Would love to merge this asap

Sorry for being slow to check in on this – I've been super busy. Let me know if I can help with this in any way?

**Comment by Jeadie** [comment]
No problem, I understand early startup days can be intense. I often get transient or odd errors in various algorithm specific tests. Will just want to make sure either a) you know the issue with a specific test, or b) you accept they won't be working in this PR run and we can work on improving/fixing post-merge.

**Comment by erikbern** [comment]
Let's merge this – I think it puts it in a better place for further work! Thanks a lot for this contribution!!!

**Pull request merged**